File: test/language/array_types/cpp98/AutoArrayRecursionTest.cpp
#include "gtest/gtest.h"

#include "array_types/auto_array_recursion/AutoArrayRecursion.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

namespace array_types
{
namespace auto_array_recursion
{

class AutoArrayRecursionTest : public ::testing::Test
{
protected:
    void fillAutoArrayRecursion(AutoArrayRecursion& autoArrayRecursion, size_t numElements)
    {
        autoArrayRecursion.setId(0);

        zserio::ObjectArray<AutoArrayRecursion>& autoArray = autoArrayRecursion.getAutoArrayRecursion();
        for (size_t i = 1; i <= numElements; ++i)
        {
            AutoArrayRecursion element;
            element.setId(static_cast<uint8_t>(i));
            autoArray.push_back(element);
        }
    }

    size_t getAutoArrayRecursionBitSizeOf(size_t numElements)
    {
        return 8 + 8 + numElements * (8 + 8);
    }

    void writeAutoArrayRecursionToByteArray(zserio::BitStreamWriter& writer, size_t numElements)
    {
        writer.writeBits(0, 8);
        writer.writeVarUInt64(static_cast<uint64_t>(numElements));
        for (size_t i = 1; i <= numElements; ++i)
        {
            writer.writeBits(static_cast<uint32_t>(i), 8);
            writer.writeVarUInt64(0);
        }
    }

    void checkAutoArrayRecursion(const AutoArrayRecursion& autoArrayRecursion, size_t numElements)
    {
        ASSERT_EQ(0, autoArrayRecursion.getId());
        const zserio::ObjectArray<AutoArrayRecursion>& autoArray = autoArrayRecursion.getAutoArrayRecursion();
        ASSERT_EQ(numElements, autoArray.size());
        for (size_t i = 1; i <= numElements; ++i)
        {
            const AutoArrayRecursion& element = autoArray.at(i - 1);
            ASSERT_EQ(i, element.getId());
            ASSERT_EQ(0, element.getAutoArrayRecursion().size());
        }
    }

    void checkBitSizeOf(size_t numElements)
    {
        AutoArrayRecursion autoArrayRecursion;
        fillAutoArrayRecursion(autoArrayRecursion, numElements);

        const size_t bitPosition = 2;
        const size_t autoArrayBitSize = getAutoArrayRecursionBitSizeOf(numElements);
        ASSERT_EQ(autoArrayBitSize, autoArrayRecursion.bitSizeOf(bitPosition));
    }

    void checkInitializeOffsets(size_t numElements)
    {
        AutoArrayRecursion autoArrayRecursion;
        fillAutoArrayRecursion(autoArrayRecursion, numElements);

        const size_t bitPosition = 2;
        const size_t expectedEndBitPosition = bitPosition + getAutoArrayRecursionBitSizeOf(numElements);
        ASSERT_EQ(expectedEndBitPosition, autoArrayRecursion.initializeOffsets(bitPosition));
    }

    void checkRead(size_t numElements)
    {
        zserio::BitStreamWriter writer;
        writeAutoArrayRecursionToByteArray(writer, numElements);
        size_t writeBufferByteSize;
        const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
        zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);

        AutoArrayRecursion autoArrayRecursion(reader);
        checkAutoArrayRecursion(autoArrayRecursion, numElements);
    }

    void checkWrite(size_t numElements)
    {
        AutoArrayRecursion autoArrayRecursion;
        fillAutoArrayRecursion(autoArrayRecursion, numElements);

        zserio::BitStreamWriter writer;
        autoArrayRecursion.write(writer);
        size_t writeBufferByteSize;
        const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
        zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);

        AutoArrayRecursion readAutoArrayRecursion(reader);
        checkAutoArrayRecursion(readAutoArrayRecursion, numElements);
    }

    static const size_t AUTO_ARRAY_LENGTH1;
    static const size_t AUTO_ARRAY_LENGTH2;
};

const size_t AutoArrayRecursionTest::AUTO_ARRAY_LENGTH1 = 5;
const size_t AutoArrayRecursionTest::AUTO_ARRAY_LENGTH2 = 10;

TEST_F(AutoArrayRecursionTest, bitSizeOfLength1)
{
    checkBitSizeOf(AUTO_ARRAY_LENGTH1);
}

TEST_F(AutoArrayRecursionTest, bitSizeOfLength2)
{
    checkBitSizeOf(AUTO_ARRAY_LENGTH2);
}

TEST_F(AutoArrayRecursionTest, initializeOffsetsLength1)
{
    checkInitializeOffsets(AUTO_ARRAY_LENGTH1);
}

TEST_F(AutoArrayRecursionTest, initializeOffsetsLength2)
{
    checkInitializeOffsets(AUTO_ARRAY_LENGTH2);
}

TEST_F(AutoArrayRecursionTest, readLength1)
{
    checkRead(AUTO_ARRAY_LENGTH1);
}

TEST_F(AutoArrayRecursionTest, readLength2)
{
    checkRead(AUTO_ARRAY_LENGTH2);
}

TEST_F(AutoArrayRecursionTest, writeLength1)
{
    checkWrite(AUTO_ARRAY_LENGTH1);
}

TEST_F(AutoArrayRecursionTest, writeLength2)
{
    checkWrite(AUTO_ARRAY_LENGTH2);
}

} // namespace auto_array_recursion
} // namespace array_types

File: test/language/array_types/python/AutoArrayRecursionTest.py
import unittest
import zserio

from testutils import getZserioApi

class AutoArrayRecursionTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "array_types.zs").auto_array_recursion

    def testBitSizeOfLength1(self):
        self._checkBitSizeOf(self.AUTO_ARRAY_LENGTH1)

    def testBitSizeOfLength2(self):
        self._checkBitSizeOf(self.AUTO_ARRAY_LENGTH2)

    def testInitializeOffsetsLength1(self):
        self._checkInitializeOffsets(self.AUTO_ARRAY_LENGTH1)

    def testInitializeOffsetsLength2(self):
        self._checkInitializeOffsets(self.AUTO_ARRAY_LENGTH2)

    def testReadLength1(self):
        self._checkRead(self.AUTO_ARRAY_LENGTH1)

    def testReadLength2(self):
        self._checkRead(self.AUTO_ARRAY_LENGTH2)

    def testWriteLength1(self):
        self._checkWrite(self.AUTO_ARRAY_LENGTH1)

    def testWriteLength2(self):
        self._checkWrite(self.AUTO_ARRAY_LENGTH2)

    def _checkBitSizeOf(self, numElements):
        autoArrayRecursion = self._createAutoArrayRecursion(numElements)
        bitPosition = 2
        autoArrayRecursionBitSize = 8 + 8 + numElements * (8 + 8)
        self.assertEqual(autoArrayRecursionBitSize, autoArrayRecursion.bitSizeOf(bitPosition))

    def _checkInitializeOffsets(self, numElements):
        autoArrayRecursion = self._createAutoArrayRecursion(numElements)
        bitPosition = 2
        expectedEndBitPosition = bitPosition + 8 + 8 + numElements * (8 + 8)
        self.assertEqual(expectedEndBitPosition, autoArrayRecursion.initializeOffsets(bitPosition))

    def _checkRead(self, numElements):
        writer = zserio.BitStreamWriter()
        AutoArrayRecursionTest._writeAutoArrayRecursionToStream(writer, numElements)

        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        autoArrayRecursion = self.api.AutoArrayRecursion.fromReader(reader)
        self._checkAutoArrayRecursion(autoArrayRecursion, numElements)

    def _checkWrite(self, numElements):
        autoArrayRecursion = self._createAutoArrayRecursion(numElements)
        bitBuffer = zserio.serialize(autoArrayRecursion)
        readAutoArrayRecursion = zserio.deserialize(self.api.AutoArrayRecursion, bitBuffer)
        self._checkAutoArrayRecursion(readAutoArrayRecursion, numElements)

    def _createAutoArrayRecursion(self, numElements):
        autoArray = []
        for i in range(1, numElements + 1):
            element = self.api.AutoArrayRecursion(i, [])
            autoArray.append(element)

        return self.api.AutoArrayRecursion(0, autoArray)

    def _checkAutoArrayRecursion(self, autoArrayRecursion, numElements):
        self.assertEqual(0, autoArrayRecursion.id)
        autoArray = autoArrayRecursion.auto_array_recursion
        self.assertEqual(numElements, len(autoArray))
        for i in range(1, numElements + 1):
            element = autoArray[i - 1]
            self.assertEqual(i, element.id)
            self.assertEqual(0, len(element.auto_array_recursion))

    @staticmethod
    def _writeAutoArrayRecursionToStream(writer, numElements):
        writer.write_bits(0, 8)
        writer.write_varsize(numElements)
        for i in range(1, numElements + 1):
            writer.write_bits(i, 8)
            writer.write_varsize(0)

    AUTO_ARRAY_LENGTH1 = 5
    AUTO_ARRAY_LENGTH2 = 10
